File: volume_1/chapter_5/llamaindex/llm_memory_persist_gateway.py
"""LLM Memory History Gateway - LlamaIndex with persistent session memory."""

import os
import sys
from pathlib import Path
from typing import Dict

CHAPTER_4_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "chapter_4"))
sys.path.append(CHAPTER_4_ROOT)
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from llama_index.core.memory import ChatMemoryBuffer
from llama_index.core.storage.chat_store import SimpleChatStore

from llm_memory_gateway import LlamaIndexLLMManager as InMemoryLlamaIndexLLMManager
from utils import interactive_cli


class LlamaIndexLLMManager(InMemoryLlamaIndexLLMManager):
    """Persistent-memory variant that reuses the in-memory manager behavior."""

    def __init__(self, memory_enabled: bool = True):
        super().__init__(memory_enabled=memory_enabled)
        self.framework = "LlamaIndex+History"
        self.chat_stores: Dict[tuple, SimpleChatStore] = {}

    def _session_file_path(self, provider: str, session_id: str) -> Path:
        sessions_dir = Path(__file__).resolve().parent / "sessions"
        sessions_dir.mkdir(parents=True, exist_ok=True)
        return sessions_dir / f"{provider}__{session_id}.json"

    def _session_store_key(self, provider: str, session_id: str) -> str:
        return f"{provider}__{session_id}"

    def _get_chat_store(self, provider: str, session_id: str) -> SimpleChatStore:
        key = (provider, session_id)
        if key in self.chat_stores:
            return self.chat_stores[key]

        path = self._session_file_path(provider, session_id)
        if path.exists():
            chat_store = SimpleChatStore.from_persist_path(str(path))
        else:
            chat_store = SimpleChatStore()
        self.chat_stores[key] = chat_store
        return chat_store

    def _get_memory(self, provider: str, session_id: str) -> ChatMemoryBuffer:
        key = (provider, session_id)
        if key not in self.memories:
            chat_store = self._get_chat_store(provider, session_id)
            self.memories[key] = ChatMemoryBuffer.from_defaults(
                chat_store=chat_store,
                chat_store_key=self._session_store_key(provider, session_id),
            )
        return self.memories[key]

    def _persist_memory(self, provider: str, session_id: str):
        chat_store = self._get_chat_store(provider, session_id)
        chat_store.persist(persist_path=str(self._session_file_path(provider, session_id)))

    def ask_question(self, *args, **kwargs):
        result = super().ask_question(*args, **kwargs)
        if result.get("success") and self.memory_enabled:
            self._persist_memory(result["provider"], result.get("session_id", "default"))
        return result

    def reset_memory(self, provider: str = None, session_id: str = None) -> Dict:
        result = super().reset_memory(provider=provider, session_id=session_id)

        if result["removed_sessions"] == ["ALL"]:
            self.chat_stores.clear()
            for path in (Path(__file__).resolve().parent / "sessions").glob("*.json"):
                path.unlink(missing_ok=True)
            return result

        for key in result["removed_sessions"]:
            if isinstance(key, tuple):
                self._session_file_path(*key).unlink(missing_ok=True)
                self.chat_stores.pop(key, None)

        return result


def main():
    if len(sys.argv) > 1 and sys.argv[1] == "web":
        from web import run_web_server

        run_web_server(lambda: LlamaIndexLLMManager(memory_enabled=True))
    else:
        interactive_cli(LlamaIndexLLMManager(memory_enabled=True))


if __name__ == "__main__":
    main()

File: volume_1/chapter_5/llamaindex/llm_memory_persist_gateway.js
/**
 * LLM Memory History Gateway - LlamaIndex JS with persistent session memory.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

import { ChatMemoryBuffer, SimpleChatStore } from 'llamaindex';
import { LlamaIndexLLMManager as InMemoryLlamaIndexLLMManager } from './llm_memory_gateway.js';
import { interactiveCli } from '../../chapter_4/utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

class LlamaIndexLLMManager extends InMemoryLlamaIndexLLMManager {
    constructor(memoryEnabled = true) {
        super(memoryEnabled);
        this.framework = 'LlamaIndex+History JS';
        this.chatStores = new Map();
    }

    _sessionFilePath(provider, sessionId) {
        const sessionsDir = path.join(__dirname, 'sessions');
        fs.mkdirSync(sessionsDir, { recursive: true });
        return path.join(sessionsDir, `${provider}__${sessionId}.json`);
    }

    _sessionStoreKey(provider, sessionId) {
        return `${provider}__${sessionId}`;
    }

    _getChatStore(provider, sessionId) {
        const key = this._historyKey(provider, sessionId);
        if (!this.chatStores.has(key)) {
            const filePath = this._sessionFilePath(provider, sessionId);
            const store = fs.existsSync(filePath)
                ? SimpleChatStore.fromPersistPath(filePath)
                : new SimpleChatStore();
            this.chatStores.set(key, store);
        }
        return this.chatStores.get(key);
    }

    _getMemory(provider, sessionId) {
        const key = this._historyKey(provider, sessionId);
        if (!this.memories.has(key)) {
            const chatStore = this._getChatStore(provider, sessionId);
            const memory = ChatMemoryBuffer.fromDefaults({
                chatStore,
                chatStoreKey: this._sessionStoreKey(provider, sessionId),
            });
            this.memories.set(key, memory);
        }
        return this.memories.get(key);
    }

    _persistMemory(provider, sessionId) {
        const chatStore = this._getChatStore(provider, sessionId);
        chatStore.persist(this._sessionFilePath(provider, sessionId));
    }

    async askQuestion(topic, provider = null, template = '{topic}', maxTokens = 1000, temperature = 0.7, sessionId = 'default') {
        const result = await super.askQuestion(topic, provider, template, maxTokens, temperature, sessionId);
        if (result.success && this.memoryEnabled) {
            this._persistMemory(result.provider, result.sessionId || 'default');
        }
        return result;
    }

    resetMemory(provider = null, sessionId = null) {
        const result = super.resetMemory(provider, sessionId);
        const sessionsDir = path.join(__dirname, 'sessions');

        if (result.removedSessions.length === 1 && result.removedSessions[0] === 'ALL') {
            this.chatStores.clear();
            if (fs.existsSync(sessionsDir)) {
                for (const fileName of fs.readdirSync(sessionsDir)) {
                    if (fileName.endsWith('.json')) {
                        fs.unlinkSync(path.join(sessionsDir, fileName));
                    }
                }
            }
            return result;
        }

        for (const key of result.removedSessions) {
            if (Array.isArray(key)) {
                const filePath = this._sessionFilePath(key[0], key[1]);
                if (fs.existsSync(filePath)) {
                    fs.unlinkSync(filePath);
                }
                this.chatStores.delete(this._historyKey(key[0], key[1]));
            }
        }

        return result;
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length > 0 && args[0] === 'web') {
        const { runWebServer } = await import('../../chapter_4/web.js');
        await runWebServer(() => new LlamaIndexLLMManager(true));
    } else {
        const manager = new LlamaIndexLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LlamaIndexLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}
